File: web/src/lib/types/network/wireguard.ts
import { z } from 'zod/v4';

export const WireGuardServerPeerSchema = z.object({
    id: z.number().int(),
    name: z.string(),
    enabled: z.boolean(),
    wireguardServerId: z.number().int(),
    privateKey: z.string(),
    publicKey: z.string(),
    preSharedKey: z.string(),
    clientIPs: z.array(z.string()),
    routableIPs: z.array(z.string()).nullish().transform((value) => value ?? []),
    routeIPs: z.boolean(),
    persistentKeepalive: z.boolean(),
    lastHandshake: z.string(),
    rx: z.number().int().nonnegative(),
    tx: z.number().int().nonnegative(),
    createdAt: z.string(),
    updatedAt: z.string()
});

export const WireGuardServerSchema = z.object({
    id: z.number().int(),
    enabled: z.boolean(),
    port: z.number().int(),
    addresses: z.array(z.string()),
    allowWireGuardPort: z.boolean().nullish().transform((value) => value ?? false),
    masqueradeIPv4Interface: z.string().nullish().transform((value) => value ?? ''),
    masqueradeIPv6Interface: z.string().nullish().transform((value) => value ?? ''),
    privateKey: z.string(),
    publicKey: z.string(),
    peers: z.array(WireGuardServerPeerSchema),
    mtu: z.number().int(),
    metric: z.number().int(),
    rx: z.number().int().nonnegative(),
    tx: z.number().int().nonnegative(),
    uptime: z.number().int().nonnegative(),
    lastHandshake: z.string(),
    restartedAt: z.string(),
    createdAt: z.string(),
    updatedAt: z.string()
});

export const WireGuardClientSchema = z.object({
    id: z.number().int(),
    enabled: z.boolean(),
    name: z.string(),
    endpointHost: z.string(),
    endpointPort: z.number().int(),
    listenPort: z.number().int(),
    privateKey: z.string(),
    publicKey: z.string(),
    peerPublicKey: z.string(),
    preSharedKey: z.string(),
    allowedIPs: z.array(z.string()),
    addresses: z.array(z.string()),
    routeAllowedIPs: z.boolean(),
    mtu: z.number().int(),
    metric: z.number().int(),
    fib: z.number().int(),
    persistentKeepalive: z.boolean(),
    rx: z.number().int().nonnegative(),
    tx: z.number().int().nonnegative(),
    uptime: z.number().int().nonnegative(),
    lastHandshake: z.string(),
    restartedAt: z.string(),
    createdAt: z.string(),
    updatedAt: z.string()
});

export const WireGuardClientStatusSchema = z.enum(['active', 'idle', 'disconnected', 'disabled']);

export type WireGuardServer = z.infer<typeof WireGuardServerSchema>;
export type WireGuardServerPeer = z.infer<typeof WireGuardServerPeerSchema>;
export type WireGuardClient = z.infer<typeof WireGuardClientSchema>;
export type WireGuardClientStatus = z.infer<typeof WireGuardClientStatusSchema>;

export function wireGuardClientStatus(client: WireGuardClient): WireGuardClientStatus {
    if (!client.enabled) {
        return 'disabled';
    }

    const handshake = Date.parse(client.lastHandshake);
    if (Number.isNaN(handshake) || handshake <= Date.parse('2001-01-01T00:00:00Z')) {
        return 'idle';
    }

    const ageSeconds = (Date.now() - handshake) / 1000;
    if (ageSeconds > 180) {
        return 'disconnected';
    }

    return 'active';
}

File: internal/db/models/network/wireguard.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package networkModels

import "time"

type WireGuardServer struct {
	ID uint `json:"id" gorm:"primaryKey"`

	Enabled bool `json:"enabled" gorm:"default:true"`

	Port      uint     `json:"port"`
	Addresses []string `json:"addresses" gorm:"serializer:json;type:json"`

	AllowWireGuardPort      bool   `json:"allowWireGuardPort" gorm:"not null;default:false"`
	MasqueradeIPv4Interface string `json:"masqueradeIPv4Interface"`
	MasqueradeIPv6Interface string `json:"masqueradeIPv6Interface"`

	PrivateKey string `json:"privateKey"`
	PublicKey  string `json:"publicKey"`

	Peers []WireGuardServerPeer `json:"peers" gorm:"foreignKey:WireGuardServerID"`

	LastKernelRX uint64 `json:"-"`
	LastKernelTX uint64 `json:"-"`
	MTU          uint   `json:"mtu"`
	Metric       uint   `json:"metric"`

	RX uint64 `json:"rx"`
	TX uint64 `json:"tx"`

	Uptime        uint64    `json:"uptime"`
	LastHandshake time.Time `json:"lastHandshake"`
	RestartedAt   time.Time `json:"restartedAt"`
	CreatedAt     time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt     time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}

type WireGuardServerPeer struct {
	ID uint `json:"id" gorm:"primaryKey"`

	Name              string `json:"name"`
	Enabled           bool   `json:"enabled" gorm:"default:true"`
	WireGuardServerID uint   `json:"wireguardServerId"`

	PrivateKey   string `json:"privateKey"`
	PublicKey    string `json:"publicKey"`
	PreSharedKey string `json:"preSharedKey"`

	ClientIPs           []string `json:"clientIPs" gorm:"serializer:json;type:json"`
	RoutableIPs         []string `json:"routableIPs" gorm:"serializer:json;type:json"`
	RouteIPs            bool     `json:"routeIPs"`
	PersistentKeepalive bool     `json:"persistentKeepalive"`

	LastHandshake time.Time `json:"lastHandshake"`

	RX           uint64 `json:"rx"`
	TX           uint64 `json:"tx"`
	LastKernelRX uint64 `json:"-"`
	LastKernelTX uint64 `json:"-"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}

type WireGuardClient struct {
	ID uint `json:"id" gorm:"primaryKey"`

	Enabled bool `json:"enabled" gorm:"default:true"`

	Name         string `json:"name" gorm:"uniqueIndex"`
	EndpointHost string `json:"endpointHost"`
	EndpointPort uint   `json:"endpointPort"`
	ListenPort   uint   `json:"listenPort"`

	PrivateKey    string `json:"privateKey"`
	PublicKey     string `json:"publicKey"`
	PeerPublicKey string `json:"peerPublicKey"`
	PreSharedKey  string `json:"preSharedKey"`

	AllowedIPs      []string `json:"allowedIPs" gorm:"serializer:json;type:json"`
	Addresses       []string `json:"addresses" gorm:"serializer:json;type:json"`
	RouteAllowedIPs bool     `json:"routeAllowedIPs"`

	MTU                 uint `json:"mtu"`
	Metric              uint `json:"metric"`
	FIB                 uint `json:"fib"`
	PersistentKeepalive bool `json:"persistentKeepalive"`

	RX           uint64 `json:"rx"`
	TX           uint64 `json:"tx"`
	KernelLastRX uint64 `json:"-"`
	KernelLastTX uint64 `json:"-"`

	Uptime        uint64    `json:"uptime"`
	LastHandshake time.Time `json:"lastHandshake"`
	RestartedAt   time.Time `json:"restartedAt"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}
